packable/packable/src/packable/map.rs

```diff
@@ -0,0 +1,239 @@
+        match self {
+            Self::DuplicateKey(_) => write!(f, "duplicate key in map"),
+            Self::Key(err) => write!(f, "cannot unpack key: {}", err),
+            Self::Value(err) => write!(f, "cannot unpack value: {}", err),
+            Self::Prefix(err) => write!(f, "cannot unpack prefix: {}", err),
+        }
+    }
+}
+
+/// Error type raised when a semantic error occurs while unpacking an ordered map.
+pub enum UnpackOrderedMapError<K, KE, VE, P> {
+    /// A map error.
+    Map(UnpackMapError<K, KE, VE, P>),
+    /// An unordered map.
+    Unordered,
+}
+
+impl<K, KE, VE, P> From<UnpackMapError<K, KE, VE, P>> for UnpackOrderedMapError<K, KE, VE, P> {
+    fn from(value: UnpackMapError<K, KE, VE, P>) -> Self {
+        Self::Map(value)
+    }
+}
+
+impl<K, KE: fmt::Debug, VE: fmt::Debug, P: fmt::Debug> fmt::Debug for UnpackOrderedMapError<K, KE, VE, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Map(arg0) => f.debug_tuple("Map").field(arg0).finish(),
+            Self::Unordered => f.debug_tuple("Unordered").finish(),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<K, KE, VE, P> std::error::Error for UnpackOrderedMapError<K, KE, VE, P>
+where
+    KE: std::error::Error,
+    VE: std::error::Error,
+    P: std::error::Error,
+{
+}
+
+impl<K, KE, VE, P> From<Infallible> for UnpackOrderedMapError<K, KE, VE, P> {
+    fn from(err: Infallible) -> Self {
+        match err {}
+    }
+}
+
+impl<K, KE: fmt::Display, VE: fmt::Display, P: fmt::Display> fmt::Display for UnpackOrderedMapError<K, KE, VE, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Map(s) => s.fmt(f),
+            Self::Unordered => write!(f, "unordered map"),
+        }
+    }
+}
+
+#[cfg(all(feature = "std", feature = "usize"))]
+impl<K: Packable + Ord + Hash, V: Packable> Packable for HashMap<K, V>
+where
+    V::UnpackVisitor: Borrow<K::UnpackVisitor>,
+{
+    type UnpackError = UnpackMapError<K, K::UnpackError, V::UnpackError, <usize as Packable>::UnpackError>;
+    type UnpackVisitor = V::UnpackVisitor;
+
+    #[inline]
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        // This cast is fine because we know `usize` is not larger than `64` bits.
+        (self.len() as u64).pack(packer)?;
+
+        for (k, v) in self.iter() {
+            k.pack(packer)?;
+            v.pack(packer)?;
+        }
+
+        Ok(())
+    }
+
+    #[inline]
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+        visitor: &Self::UnpackVisitor,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        use crate::error::UnpackErrorExt;
+
+        let len = u64::unpack::<_, VERIFY>(unpacker, &())
+            .coerce()?
+            .try_into()
+            .map_err(|err| UnpackError::Packable(UnpackMapError::Prefix(err)))?;
+
+        let mut map = HashMap::<K, V>::new();
```

Comment:
have you considered `with_capacity` here?

Reply:
done